packages/labs/compiler/src/lib/ast-fragments.ts

```diff
@@ -7,14 +7,175 @@
 import ts from 'typescript';
 
 export const PartType = {
+  ATTRIBUTE: 1,
   CHILD: 2,
+  PROPERTY: 3,
+  BOOLEAN_ATTRIBUTE: 4,
+  EVENT: 5,
+  ELEMENT: 6,
+  COMMENT_PART: 7,
 } as const;
 
 export type PartType = (typeof PartType)[keyof typeof PartType];
 
-export type TemplatePart = {
-  type: typeof PartType.CHILD;
-  index: number;
+export type TemplatePart =
+  | {
+      type:
+        | typeof PartType.CHILD
+        | typeof PartType.COMMENT_PART
+        | typeof PartType.ELEMENT;
+      index: number;
+    }
+  | {
+      type: typeof PartType.ATTRIBUTE;
+      index: number;
+      name: string;
+      strings: Array<string>;
+      ctorType: PartType;
+    };
+
+const attributePartConstructors = {
+  [PartType.ATTRIBUTE]: 'AttributePart',
+  [PartType.PROPERTY]: 'PropertyPart',
+  [PartType.BOOLEAN_ATTRIBUTE]: 'BooleanAttributePart',
+  [PartType.EVENT]: 'EventPart',
+} as const;
+
+/**
+ * Mapping of part constructors to a unique identifier alias. If undefined, the
+ * constructor will not be imported.
+ */
+export interface AttributePartConstructorAliases {
+  AttributePart?: ts.Identifier;
+  PropertyPart?: ts.Identifier;
+  BooleanAttributePart?: ts.Identifier;
+  EventPart?: ts.Identifier;
+}
+
+/**
+ * An array of names for the attribute constructors that are exported from the
+ * module `"lit-html/private-ssr-support.js"` via the `_$LH` object.
+ */
+const attributePartConstructorNames = Object.values(attributePartConstructors);
+
+/**
+ * Creates the statement:
+ *
+ * ```ts
+ * import * as <namespace> from "<moduleSpecifier>"
+ * ```
+ */
+const createImportNamespaceDeclaration = (
+  factory: ts.NodeFactory,
+  namespace: ts.Identifier,
+  moduleSpecifier: ts.StringLiteral
+) =>
+  factory.createImportDeclaration(
+    undefined,
+    factory.createImportClause(
+      false,
+      undefined,
+      factory.createNamespaceImport(namespace)
+    ),
+    moduleSpecifier,
+    undefined
+  );
+
+/**
+ * Add Parts constructors import. To avoid identifier collisions, each part
+ * constructor is mapped to a unique identifier that must be provided. If no
+ * identifier is provided for a part, then the part will not be emitted.
```

Comment:
> If no identifier is provided for a part, then the part will not be emitted.

what does that mean? should we instead throw in that case?

Reply:
This means that if a certain part constructor does not have a unique identifier provided, we ignore it and don't pull it off `_$LH`.

This prevents compiled files containing every part constructor. We only import what is used.

Reply:
Oh, I see. I initially read it like "in [this case], [this parameter] is ignored", which set off my warning bells, but it's more like "add imports as requested by the parameters". The only things that are ignored are the values which aren't requested, i.e. that aren't in the attributePartConstructorNameMap

Reply:
Exactly! I'll reword the comment to make that clearer

Reply:
Reworded in https://github.com/lit/lit/pull/4056/commits/54a40ce01e717c67e23371402d9b0fb0fbded930